modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationListenerHolder.java

```diff
@@ -37,12 +37,15 @@
      * @see ConfigurationListenerHolder#listeners
      */
     void addListener(L listener, long notificationNumber) {
-        containers.add(new Container<>(listener, notificationNumber));
+        containers.add(new Container<>(listener, notificationNumber + 1));
     }
 
     /**
      * Removes the listener.
      *
+     * <p>NOTE: This method introduces unpredictable behavior at the moment, because the final behavior of this method is unclear.
```

Comment:
I think that such documentation on "public" API would be better. No one will read this class except for you and I, probably.
We should also explicitly state all guarantees that we do or don't give, like the order of execution on the same node and the ability to add listeners inside of other listeners with expected behavior. Do we have such documentation?